File: src/Runtime/CPP/Dataset/HDF5Dataset.cpp
#include "HDF5Dataset.hpp"

#include <algorithm>
#include <limits>
#include <numeric>
#include <stdexcept>
#include <utility>

namespace PHP2xAI::Runtime::CPP
{
	HDF5Dataset::HDF5Dataset(
		std::string filename,
		std::size_t batchSize,
		std::string xField,
		std::string yField,
		std::uint32_t seed)
		: dataset_(PHP2XAIHDF5::open(filename)),
		  filename_(std::move(filename)),
		  batchSize_(batchSize),
		  xField_(std::move(xField)),
		  yField_(std::move(yField)),
		  sampleCount_(0),
		  xElementsPerSample_(0),
		  yElementsPerSample_(0),
		  rng_(seed)
	{
		if (batchSize_ == 0)
			throw std::runtime_error("batchSize must be > 0");

		xMetadata_ = dataset_->fieldMetadata(xField_);
		yMetadata_ = dataset_->fieldMetadata(yField_);
		xElementsPerSample_ = elementCount(xMetadata_.shape);
		yElementsPerSample_ = elementCount(yMetadata_.shape);
		sampleCount_ = dataset_->count();

		if (sampleCount_ < static_cast<std::int64_t>(batchSize_))
			throw std::runtime_error("HDF5 dataset has no complete batches");

		sampleOrder_.resize(static_cast<std::size_t>(sampleCount_));
		std::iota(sampleOrder_.begin(), sampleOrder_.end(), std::int64_t{0});
	}

	std::size_t HDF5Dataset::numBatches() const
	{
		return static_cast<std::size_t>(sampleCount_) / batchSize_;
	}

	std::int64_t HDF5Dataset::sampleCount() const
	{
		return sampleCount_;
	}

	std::size_t HDF5Dataset::batchSize() const
	{
		return batchSize_;
	}

	std::string HDF5Dataset::getType() const
	{
		return "HDF5";
	}

	void HDF5Dataset::shuffleEpoch()
	{
		std::shuffle(sampleOrder_.begin(), sampleOrder_.end(), rng_);
		resetEpoch();
	}

	void HDF5Dataset::resetEpoch()
	{
		batchPosition_ = 0;
		currentIndices_.clear();
	}

	bool HDF5Dataset::nextBatch()
	{
		if (!currentIndices_.empty())
			return true;
		if (batchPosition_ >= numBatches())
			return false;

		const std::size_t first = batchPosition_ * batchSize_;
		currentIndices_.assign(
			sampleOrder_.begin() + static_cast<std::ptrdiff_t>(first),
			sampleOrder_.begin() + static_cast<std::ptrdiff_t>(first + batchSize_));
		return true;
	}

	void HDF5Dataset::pack(std::vector<float>& xPacked, std::vector<float>& yPacked)
	{
		if (currentIndices_.empty())
			throw std::runtime_error("Call nextBatch() before pack()");

		readFieldAsFloat(xField_, xMetadata_.dtype, xElementsPerSample_, currentIndices_, xPacked);
		readFieldAsFloat(yField_, yMetadata_.dtype, yElementsPerSample_, currentIndices_, yPacked);

		++batchPosition_;
		currentIndices_.clear();
	}

	std::size_t HDF5Dataset::elementCount(const std::vector<std::int64_t>& shape)
	{
		if (shape.empty())
			throw std::runtime_error("HDF5 field shape cannot be empty");

		std::size_t count = 1;
		for (const std::int64_t dimension : shape)
		{
			if (dimension <= 0)
				throw std::runtime_error("HDF5 field dimensions must be positive");
			const auto size = static_cast<std::size_t>(dimension);
			if (count > std::numeric_limits<std::size_t>::max() / size)
				throw std::runtime_error("HDF5 field shape is too large");
			count *= size;
		}
		return count;
	}

	void HDF5Dataset::readFieldAsFloat(
		const std::string& field,
		PHP2XAIHDF5::DType dtype,
		std::size_t elementsPerSample,
		const std::vector<std::int64_t>& indices,
		std::vector<float>& output) const
	{
		if (indices.size() > std::numeric_limits<std::size_t>::max() / elementsPerSample)
			throw std::runtime_error("HDF5 output buffer is too large");
		const std::size_t valueCount = indices.size() * elementsPerSample;

		auto readAndConvert = [&](auto value) {
			using Value = decltype(value);
			std::vector<Value> values(valueCount);
			dataset_->readIndices(field, indices, values.data());
			output.resize(valueCount);
			std::transform(values.begin(), values.end(), output.begin(),
				[](Value item) { return static_cast<float>(item); });
		};

		switch (dtype)
		{
			case PHP2XAIHDF5::FLOAT32:
				readAndConvert(float{});
				break;
			case PHP2XAIHDF5::FLOAT64:
				readAndConvert(double{});
				break;
			case PHP2XAIHDF5::INT32:
				readAndConvert(std::int32_t{});
				break;
			case PHP2XAIHDF5::INT64:
				readAndConvert(std::int64_t{});
				break;
			default:
				throw std::runtime_error("Unsupported HDF5 dtype");
		}
	}
}

File: src/Runtime/PHP/Datasets/HDF5Dataset.php
<?php

namespace PHP2xAI\Runtime\PHP\Datasets;

use PHP2xAI\Tensor\Tensor;
use RuntimeException;

/** Batch iterator over feature and target fields stored in an HDF5 file. */
class HDF5Dataset extends BatchDataset implements \IteratorAggregate
{
	private ?PHP2XAIHDF5 $dataset;
	private string $path;
	private int $batchSize;
	private string $xField;
	private string $yField;
	private int $sampleCount;
	private array $sampleOrder = [];
	private int $batchPosition = 0;
	private ?array $currentIndices = null;
	private array $xMetadata;
	private array $yMetadata;
	private ?Tensor $xPlaceholder = null;
	private ?Tensor $yPlaceholder = null;

	public function __construct(
		string $filename,
		int $batchSize,
		string $xField = 'x',
		string $yField = 'y'
	) {
		if ($batchSize <= 0)
			throw new \InvalidArgumentException('batchSize must be > 0');

		$this->path = $filename;
		$this->dataset = PHP2XAIHDF5::open($filename);
		$this->batchSize = $batchSize;
		$this->xField = $xField;
		$this->yField = $yField;
		$this->xMetadata = $this->dataset->fieldMetadata($xField);
		$this->yMetadata = $this->dataset->fieldMetadata($yField);
		$this->sampleCount = $this->dataset->count();

		if ($this->sampleCount < $this->batchSize)
			throw new RuntimeException('HDF5 dataset has no complete batches');

		$this->sampleOrder = range(0, $this->sampleCount - 1);
	}

	public function getBatchSize(): int
	{
		return $this->batchSize;
	}

	public function getPath(): string
	{
		return $this->path;
	}

	public function getType(): string
	{
		return 'HDF5';
	}

	public function resetEpoch(): void
	{
		$this->batchPosition = 0;
		$this->currentIndices = null;
	}

	public function shuffleEpoch(?int $seed = null): void
	{
		if ($seed !== null)
			mt_srand($seed);
		shuffle($this->sampleOrder);
	}

	public function nextBatch(): bool
	{
		if ($this->currentIndices !== null)
			return true;
		if ($this->batchPosition >= intdiv($this->sampleCount, $this->batchSize))
			return false;

		$firstPosition = $this->batchPosition * $this->batchSize;
		$this->currentIndices = array_slice($this->sampleOrder, $firstPosition, $this->batchSize);
		return true;
	}

	/** Return [xRowMajor, yRowMajor] for the current batch. */
	public function pack(): array
	{
		if ($this->currentIndices === null)
			throw new RuntimeException('Call nextBatch() before pack()');
		$this->ensureOpen();

		$indices = $this->currentIndices;
		$x = $this->dataset->readIndices($this->xField, $indices);
		$y = $this->dataset->readIndices($this->yField, $indices);

		$this->batchPosition++;
		$this->currentIndices = null;

		return [$x, $y];
	}

	/** Iterate batches, each yielding individual [x, y] row-major samples. */
	public function getIterator(): \Traversable
	{
		$numBatches = intdiv($this->sampleCount, $this->batchSize);
		for ($batch = 0; $batch < $numBatches; $batch++) {
			$first = $batch * $this->batchSize;
			$indices = array_slice($this->sampleOrder, $first, $this->batchSize);
			yield $this->iterateBatch($indices);
		}
	}

	/** Yield each sample separately, like StreamFileDataset's iterator. */
	private function iterateBatch(array $indices): \Generator
	{
		$this->ensureOpen();
		$x = $this->dataset->readIndices($this->xField, $indices);
		$y = $this->dataset->readIndices($this->yField, $indices);
		$xSize = self::elementCount($this->xMetadata['shape']);
		$ySize = self::elementCount($this->yMetadata['shape']);

		for ($i = 0; $i < count($indices); $i++) {
			yield [
				array_slice($x, $i * $xSize, $xSize),
				array_slice($y, $i * $ySize, $ySize),
			];
		}
	}

	public function initPlaceholders(bool $train = true): void
	{
		$batch = $train ? $this->batchSize : 1;
		$xShape = array_merge([$batch], $this->xMetadata['shape']);
		$ySampleElements = self::elementCount($this->yMetadata['shape']);
		$yShape = $ySampleElements === 1
			? [$batch]
			: array_merge([$batch], $this->yMetadata['shape']);

		$this->xPlaceholder = Tensor::zeros($xShape, 'x');
		$this->yPlaceholder = Tensor::zeros($yShape, 'y');
	}

	public function getPlaceholders(): array
	{
		return ['x' => $this->xPlaceholder, 'y' => $this->yPlaceholder];
	}

	public function getXPlaceholder(): ?Tensor
	{
		return $this->xPlaceholder;
	}

	public function getYPlaceholder(): ?Tensor
	{
		return $this->yPlaceholder;
	}

	/** Close the HDF5 handle so another process can open the file. */
	public function close(): void
	{
		if ($this->dataset !== null) {
			$this->dataset->destroy();
			$this->dataset = null;
		}
	}

	private function ensureOpen(): void
	{
		if ($this->dataset === null)
			$this->dataset = PHP2XAIHDF5::open($this->path);
	}

	private static function elementCount(array $shape): int
	{
		$count = 1;
		foreach ($shape as $dimension)
			$count *= $dimension;
		return $count;
	}
}
